File: app/src/androidTest/java/ScreenshotTestRule.java
import android.app.Activity;
import android.graphics.Bitmap;
import android.view.View;

import org.junit.rules.MethodRule;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.Statement;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import static android.support.test.internal.runner.junit4.statement.UiThreadStatement.runOnUiThread;

class ScreenshotTestRule implements MethodRule {

    private static final String SCREENSHOTS_PATH = "test-screenshots";
    static final String SCREENSHOTS_PATH_SHOWCASE = "test-screenshots/showcase";

    public Statement apply(final Statement statement, final FrameworkMethod frameworkMethod, final Object o) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                try {
                    statement.evaluate();
                } catch (Throwable t) {
                    captureScreenshot(frameworkMethod.getName());
                    throw t;
                }
            }

            void captureScreenshot(String fileName) {
                takeScreenshot(fileName, WtdTest.getActivityInstance(), SCREENSHOTS_PATH);
            }
        };
    }

    static void takeScreenshot(String name, Activity activity, String path)
    {
        View scrView = activity.getWindow().getDecorView().getRootView();

        new Thread() {
            public void run() {
                try {
                    runOnUiThread(() -> {
                        scrView.setDrawingCacheEnabled(true);
                        Bitmap bitmap = Bitmap.createBitmap(scrView.getDrawingCache());
                        scrView.setDrawingCacheEnabled(false);

                        OutputStream out = null;
                        File imagePath = new File(activity.getFilesDir(), path);
                        imagePath.mkdirs();
                        File imageFile = new File(imagePath, name + ".png");

                        try {
                            out = new FileOutputStream(imageFile);
                            bitmap.compress(Bitmap.CompressFormat.PNG, 90, out);
                            out.flush();
                        } catch (FileNotFoundException e) {
                            System.err.println(e.getMessage());
                        } catch (IOException e) {
                            System.err.println(e.getMessage());
                        } finally {
                            try {
                                if (out != null) {
                                    out.close();
                                }
                            } catch (Exception e) {
                                System.err.println(e.getMessage());
                            }
                        }
                    });
                } catch (Throwable throwable) {
                    throwable.printStackTrace();
                }
            }
        }.run();
    }
}

File: app/src/androidTest/java/ScreenshotTestRule.kt


import android.app.Activity
import android.graphics.Bitmap
import android.support.test.internal.runner.junit4.statement.UiThreadStatement.runOnUiThread
import org.junit.rules.MethodRule
import org.junit.runners.model.FrameworkMethod
import org.junit.runners.model.Statement
import java.io.*

class ScreenshotTestRule : MethodRule {

    override fun apply(statement: Statement, frameworkMethod: FrameworkMethod, o: Any): Statement {
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                try {
                    statement.evaluate()
                } catch (t: Throwable) {
                    captureScreenshot(frameworkMethod.name)
                    throw t
                }

            }

            fun captureScreenshot(fileName: String) {
                takeScreenshot(fileName, WtdTest.activityInstance, SCREENSHOTS_PATH)
            }
        }
    }

    companion object {

        private val SCREENSHOTS_PATH = "test-screenshots"
        val SCREENSHOTS_PATH_SHOWCASE = "test-screenshots/showcase"

        fun takeScreenshot(name: String, activity: Activity, path: String) {
            val scrView = activity.window.decorView.rootView

            object : Thread() {
                override fun run() {
                    try {
                        runOnUiThread {
                            scrView.isDrawingCacheEnabled = true
                            val bitmap = Bitmap.createBitmap(scrView.drawingCache)
                            scrView.isDrawingCacheEnabled = false

                            var out: OutputStream? = null
                            val imagePath = File(activity.filesDir, path)
                            imagePath.mkdirs()
                            val imageFile = File(imagePath, "$name.png")

                            try {
                                out = FileOutputStream(imageFile)
                                bitmap.compress(Bitmap.CompressFormat.PNG, 90, out)
                                out.flush()
                            } catch (e: FileNotFoundException) {
                                System.err.println(e.message)
                            } catch (e: IOException) {
                                System.err.println(e.message)
                            } finally {
                                try {
                                    out?.close()
                                } catch (e: Exception) {
                                    System.err.println(e.message)
                                }

                            }
                        }
                    } catch (throwable: Throwable) {
                        throwable.printStackTrace()
                    }

                }
            }.run()
        }
    }
}